### PR TITLE
Add Evaluation / Feedback Examples

### PR DESCRIPTION
**Pull request by vowelparrot**

### CODE REVIEW

File: langchain/client/langchain.py [hunk 1]
```diff
@@ -158,8 +164,8 @@ def upload_dataframe(
         df: pd.DataFrame,
         name: str,
         description: str,
-        input_keys: List[str],
-        output_keys: List[str],
+        input_keys: Sequence[str],
```
Comment (eyurtsev) [inline langchain/client/langchain.py:167]: :heart: Immediate approval for PR!

File: langchain/client/langchain.py [hunk 2]
```diff
@@ -268,7 +271,7 @@ def list_sessions(self) -> Iterator[TracerSession]:
         raise_for_status_with_text(response)
         yield from [TracerSession(**session) for session in response.json()]
 
-    def create_dataset(self, dataset_name: str, description: str) -> Dataset:
+    def create_dataset(self, dataset_name: str, description: Optional[str]) -> Dataset:
```
Comment (eyurtsev) [inline langchain/client/langchain.py:274]:
```suggestion
    def create_dataset(self, dataset_name: str, *, description: Optional[str] = None) -> Dataset:
```

File: langchain/client/langchain.py [hunk 3]
```diff
@@ -383,6 +386,92 @@ def list_examples(
         raise_for_status_with_text(response)
         yield from [Example(**dataset) for dataset in response.json()]
 
+    def update_example(
+        self,
+        example_id: str,
```
Comment (eyurtsev) [inline langchain/client/langchain.py:391]:
```suggestion
        example_id: str,
        *,
```

File: langchain/client/langchain.py [hunk 4]
```diff
@@ -383,6 +386,92 @@ def list_examples(
...
+        inputs: Optional[Dict[str, Any]] = None,
```
Comment (eyurtsev) [inline langchain/client/langchain.py:392]:
```suggestion
        inputs: Optional[Mapping[str, Any]] = None,
```

File: langchain/client/langchain.py [hunk 5]
```diff
@@ -383,6 +386,92 @@ def list_examples(
...
+        outputs: Optional[Dict[str, Any]] = None,
```
Comment (eyurtsev) [inline langchain/client/langchain.py:393]:
```suggestion
        outputs: Optional[Mapping[str, Any]] = None,
```

File: langchain/client/langchain.py [hunk 6]
```diff
@@ -383,6 +386,92 @@ def list_examples(
...
+        )
+        response = requests.post(
+            self.api_url + "/feedback",
+            headers=self._headers,
+            data=feedback.json(),
+        )
+        raise_for_status_with_text(response)
+        return Feedback(**feedback.dict())
+
+    @retry(stop=stop_after_attempt(3), wait=wait_fixed(0.5))
```
Comment (eyurtsev) [inline langchain/client/langchain.py:444]: Are we able to specify which errors are retried? i.e., 5xx and a few selected 4xx

File: langchain/client/models.py [hunk 7]
```diff
@@ -31,12 +31,20 @@ class Example(ExampleBase):
     runs: List[Run] = Field(default_factory=list)
 
 
+class ExampleUpdate(BaseModel):
```
Comment (eyurtsev) [inline langchain/client/models.py:34]: 1) Should we be setting frozen in the config?

File: langchain/client/models.py [hunk 8]
```diff
@@ -31,12 +31,20 @@ class Example(ExampleBase):
...
+    """Update class for Example."""
+
+    dataset_id: Optional[UUID] = None
+    inputs: Optional[Dict[str, Any]] = None
```
Comment (eyurtsev) [inline langchain/client/models.py:38]:
```suggestion
    inputs: Optional[Mapping[str, Any]] = None
```

File: langchain/client/models.py [hunk 9]
```diff
@@ -31,12 +31,20 @@ class Example(ExampleBase):
...
+    outputs: Optional[Dict[str, Any]] = None
```
Comment (eyurtsev) [inline langchain/client/models.py:39]:
```suggestion
    outputs: Optional[Mapping[str, Any]] = None
```

File: langchain/client/models.py [hunk 10]
```diff
@@ -97,3 +105,57 @@ def validate_time_range(cls, values: Dict[str, Any]) -> Dict[str, Any]:
...
+    type: ClassVar[str] = "api"
+    metadata: Optional[Dict[str, Any]] = None
+
+
+class FeedbackBase(BaseModel):
+    """Feedback schema."""
+
+    created_at: datetime = Field(default_factory=datetime.utcnow)
+    """The time the feedback was created."""
+    modified_at: datetime = Field(default_factory=datetime.utcnow)
```
Comment (eyurtsev) [inline langchain/client/models.py:122]: style: i like `updated_at` better (sounds better to me when contrasted with `created_at`)

File: langchain/client/models.py [hunk 11]
```diff
@@ -97,3 +105,57 @@ def validate_time_range(cls, values: Dict[str, Any]) -> Dict[str, Any]:
...
+    """The time the feedback was last modified."""
+    run_id: UUID
+    """The associated run ID this feedback is logged for."""
+    metric_name: str
+    """The feedback metric name or type."""
+    metric_value: Optional[Union[float, str]] = None
```
Comment (eyurtsev) [inline langchain/client/models.py:128]: Should ints / bools be supported?

File: langchain/client/models.py [hunk 12]
```diff
@@ -97,3 +105,57 @@ def validate_time_range(cls, values: Dict[str, Any]) -> Dict[str, Any]:
...
+    """Score to assign the run."""
+    feedback_source: Optional[Union[APIFeedbackSource, Dict[str, Any]]] = None
```
Comment (eyurtsev) [inline langchain/client/models.py:130]:
```suggestion
    feedback_source: Optional[Union[APIFeedbackSource, Mapping[str, Any]]] = None
```

File: langchain/client/models.py [hunk 13]
```diff
@@ -97,3 +105,57 @@ def validate_time_range(cls, values: Dict[str, Any]) -> Dict[str, Any]:
...
+
+    id: UUID
+    feedback_source: Optional[Dict] = None
+    """The source of the feedback. In this case"""
+
+
+class ListFeedbackQueryParams(BaseModel):
+    """Query Params for listing feedbacks."""
+
+    run: Optional[List[UUID]] = None
```
Comment (eyurtsev) [inline langchain/client/models.py:154]:
```suggestion
    run: Optional[Sequence[UUID]] = None
```

**Review by eyurtsev — COMMENTED** [review]
Have to run to sleep! Sorry didn't manage to get through the actual PR logic only scanned briefly. (Looking good from what I can tell though.)

File: langchain/client/models.py (hunk 11 above)
Comment (vowelparrot) [reply langchain/client/models.py:128]: Ya they should!

File: langchain/client/models.py (hunk 7 above)
Comment (vowelparrot) [reply langchain/client/models.py:34]: Ya

File: langchain/client/models.py (hunk 10 above)
Comment (vowelparrot) [reply langchain/client/models.py]: Ya that may be, but we already use modified_at on other resources and I would rather be consistent with that

File: langchain/client/langchain.py (hunk 6 above)
Comment (vowelparrot) [reply langchain/client/langchain.py:448]: Ya - I think I'll update that in a subsequent PR though because our error raising isn't super specific

**Comment by vowelparrot** [comment]
Accidentally slip-fingered a reverse rebase and messed up this PR. Going to continue in #5166

**Pull request closed without merging**